resources/assets/lib/models/chat/create-announcement.ts

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import { createAnnoucement } from 'chat/chat-api';
+import UserJson from 'interfaces/user-json';
+import { route } from 'laroute';
+import { debounce } from 'lodash';
+import { action, computed, makeObservable, observable, runInAction } from 'mobx';
+import core from 'osu-core-singleton';
+
+export const inputKeys = ['description', 'message', 'name', 'users'] as const;
+type InputKey = typeof inputKeys[number];
+type Inputs = { [key in InputKey]: string };
+
+export function isInputKey(key: string): key is InputKey {
+  return (inputKeys as Readonly<string[]>).includes(key);
+}
+
+export default class CreateAnnouncement {
+  @observable busy = {
+    create: false,
+    lookupUsers: false,
+  };
+
+  debouncedLookupUsers = debounce(action(() => this.lookupUsers()), 1000);
+
+  @observable inputs: Inputs = {
+    description: '',
+    message: '',
+    name: '',
+    users: '',
+  };
+  @observable validUsers = new Map<number, UserJson>();
+
+  private uuid = osu.uuid();
+  private xhrLookupUsers?: JQueryXHR;
+
+  @computed
+  get errors() {
+    return {
+      description: !osu.present(this.inputs.description),
+      message: !osu.present(this.inputs.message),
+      name: !osu.present(this.inputs.name),
+      users: this.validUsers.size === 0
+        || osu.present(this.inputs.users.trim()), // implies invalid ids left
+    };
+  }
+
+  @computed
+  get isValid() {
+    return !Object.values(this.errors).some(Boolean);
+  }
+
+  constructor() {
+    makeObservable(this);
+  }
+
+  @action
+  clear() {
+    this.xhrLookupUsers?.abort();
+    Object.keys(this.busy).forEach((key: keyof typeof this.busy) => this.busy[key] = false);
+    Object.keys(this.inputs).forEach((key: keyof typeof this.inputs) => this.inputs[key] = '');
+    this.validUsers.clear();
+  }
+
+  @action
+  create() {
+    if (!this.isValid || this.busy.create) return;
+
+    // busy state should remain active so the same model can't be used to send multiple requests
+    // until the entire workflow is complete, including switching the channel.
+    // TODO: need a cancel or timeout or something?
+    this.busy.create = true;
+    const json = this.toJson();
+    core.dataStore.chatState.waitJoinChannelUuid = json.uuid;
+
+    createAnnoucement(json)
+      .fail(action((xhr: JQueryXHR) => {
+        osu.ajaxError(xhr);
+        this.busy.create = false;
+      }));
+  }
+
+  toJson() {
+    const { description, message, name } = this.inputs;
+
+    return {
+      channel: { description, name },
+      message,
+      target_ids: [...this.validUsers.keys()],
+      type: 'ANNOUNCE' as const,
+      uuid: this.uuid,
+    };
+  }
+
+  @action
+  updateUsers(text: string, immediate: boolean) {
+    this.busy.lookupUsers = true;
```

Comment:
emptying the input box shouldn't show the spinner. And maybe also when the change is only on whitespace.